### PR TITLE
Enhance module api

### PR DESCRIPTION
**Pull request by banditopazzo**
# Enhance module api

New features:
  - modules can have custom initialization fields
  - add methods to module trait
  - move module loops inside module manager

ref: #157

### CODE REVIEW

File: crates/modules/logger/src/lib.rs [hunk 1]
```diff
@@ -14,62 +10,76 @@ use std::{
...
+        state: &mut Self::State,
+        ctx: &ModuleContext,
+    ) -> Result<(), ModuleError> {
+        state.logger = match Logger::from_config(new_config) {
+            Ok(logr) => logr,
+            Err(logr) => {
+                ctx.raise_warning("Failed to connect to syslog".into())
+                    .await;
+                logr
             }
```
Comment (vadorovsky) [inline crates/modules/logger/src/lib.rs:59]:
```suggestion
        state.logger = match Logger::from_config(new_config).unwrap_or_else(|logr| {
            ctx.raise_warning("Failed to connect to syslog".into())
                .await;
            logr
        });
```

**Review by vadorovsky — COMMENTED** [review]
I like this change in general! Having a trait is so much cleaner than having the `Module` struct. Left some nit comments, but I will give it a second look soon.

**Review by vadorovsky — COMMENTED** [review]
I have no other comments than my previous nits, the code looks good to me.

I could re-review after we get #288 merged, but overall I like the direction and I don't think I'm going to have any not nitpicky remarks.

File: crates/modules/logger/src/lib.rs [hunk 2]
```diff
@@ -14,62 +10,76 @@ use std::{
...
+        ctx: &ModuleContext,
+    ) -> Result<Self::State, ModuleError> {
+        let logger = match Logger::from_config(config) {
+            Ok(logr) => logr,
+            Err(logr) => {
+                ctx.raise_warning("Failed to connect to syslog".into())
+                    .await;
+                logr
+            }
+        };
```
[the remark this answers is not in the document]
Comment (banditopazzo) [reply crates/modules/logger/src/lib.rs:43]: it's not possible to use `await` in the closure

File: crates/modules/logger/src/lib.rs (hunk 1 above)
Comment (banditopazzo) [reply crates/modules/logger/src/lib.rs:59]: > it's not possible to use `await` in the closure

same here

**Review by vadorovsky — APPROVED** [review]

**Comment by banditopazzo** [comment]
Closed in favor of #313 because this PR was impossible to merge due to Github Issue https://github.com/orgs/community/discussions/27521

**Pull request closed without merging**